KAFKA-12708 Rewrite org.apache.kafka.test.Microbenchmarks by JMH

brandboat opened the pull request:
related to https://issues.apache.org/jira/browse/KAFKA-12708, mostly based on https://github.com/apache/kafka/pull/10663
```bash
./jmh-benchmarks/jmh.sh org.apache.kafka.jmh.util.ConcurrentMapBenchmark
```
```
# OS: ubuntu 22.04
# JMH version: 1.37
# JDK version: OpenJDK Runtime Environment Zulu17.50+19-CA (build 17.0.11+9-LTS)
# CPU: Intel(R) Core(TM) i9-14900K

Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt      Score      Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt    5  22338.964 ± 3148.123  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt    5     13.175 ±    0.488  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt    5     16.382 ±    1.301  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt    5  19684.342 ± 2892.526  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt    5  21260.254 ±  860.984  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt    5   5436.128 ±  883.432  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt    5   5606.463 ±  962.939  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt    5  16483.967 ± 1072.444  ns/op
```

A lower value in the score indicates better performance.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-12708 Rewrite org.apache.kafka.test.Microbenchmarks by JMH

brandboat committed:
update

chia7712 commented:
@brandboat thanks for this patch. It would be nice that you add "how to reproduce the benchmark result" to the description.

chia7712 commented:
```
MapBenchmark.testCopyOnWriteMap     avgt   15  3518087.394 ± 167618.482  ns/op
MapBenchmark.testHashMap            avgt   15  3284793.930 ± 288361.593  ns/op
```
it seems `CopyOnWriteMap` is faster than `HashMap`?

TaiJuWu commented:
Could you share why `testConcurrentHashMap` is faster than `testHashMap`?
It is not reasonable for me, in my mind the work of `ConcurrentHashMap` should be large than `HashMap`. because the front need to deal with sync problem.

TaiJuWu commented:
> ```
> MapBenchmark.testCopyOnWriteMap     avgt   15  3518087.394 ± 167618.482  ns/op
> MapBenchmark.testHashMap            avgt   15  3284793.930 ± 288361.593  ns/op
> ```
> 
> it seems `CopyOnWriteMap` is faster than `HashMap`?

The benchmarl only test read part

> ```
> MapBenchmark.testCopyOnWriteMap     avgt   15  3518087.394 ± 167618.482  ns/op
> MapBenchmark.testHashMap            avgt   15  3284793.930 ± 288361.593  ns/op
> ```
> 
> it seems `CopyOnWriteMap` is faster than `HashMap`?

Maybe this is statistical error?
The benchmark only test read part but the benefit of `CopyOnWriteMap` is write part.

chia7712 commented:
Maybe we should focus on the methods used in code base.

brandboat committed:
Fix jmh

Co-authored-by:  xuanzhang gong<gongxuanzhangmelt@gmail.com>

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/java/MapBenchmark.java:
```diff
@@ -0,0 +1,79 @@
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Threads(2)
+public class MapBenchmark {
+    private final int TIMES = 1_000_000;
```
Could you please follow the JMH style? for example:
```java
    @Param({"10000"})
    private int times;

    @Param({"100"})
    private int mapSize;

    private Map<Integer, Integer> hashMap;
    private Map<Integer, Integer> concurrentHashMap;
    private Map<Integer, Integer> copyOnWriteMap;

    @Setup
    public void setup() {
        Map<Integer, Integer> mapTemplate = IntStream.range(0, mapSize).boxed()
          .collect(Collectors.toMap(i -> i, i -> i));
        hashMap = new HashMap<>(mapTemplate);
        concurrentHashMap = new ConcurrentHashMap<>(mapTemplate);
        copyOnWriteMap = new CopyOnWriteMap<>(mapTemplate);

    }
```
That gives us more control and better console output (`mapSize` and `time` are included by the output)

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/java/MapBenchmark.java:
```diff
@@ -0,0 +1,79 @@
+    @Benchmark
+    public void testHashMap(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            blackhole.consume(hashMap.get(KEYS_ARRAY[i % MAP_SIZE]));
```
Could we just loop `KEYS_ARRAY` instead of calculating remainder?

brandboat committed:
Address comments

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/java/LockBenchmark.java:
```diff
@@ -0,0 +1,68 @@
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Threads(2)
+public class LockBenchmark {
```
I don't think this JMH test is required, as that is common sense now

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/java/BasicBenchmark.java:
```diff
@@ -0,0 +1,81 @@
+@Measurement(iterations = 15)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+public class BasicBenchmark {
```
just remove this relics. This can get backed by git in the future if we miss them :)

chia7712 reviewed (commented):
@brandboat thanks for updated PR

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/java/MapBenchmark.java:
```diff
@@ -0,0 +1,94 @@
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Threads(2)
+public class MapBenchmark {
```
Could you please rename it to `ConcurrentMapBenchmark`?

brandboat committed:
Address comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

brandboat committed:
Add computeIfAbsent to benchmark

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

brandboat commented:
Modified a couple of things in the latest commit:
1. Add computeIfAbsent in each benchmark since in the code base we use both get and computeIfAbsent in CopyOnWriteMap, and we want to test the performance between CopyOnWriteMap & ConcurrentHashMap.
2. remove HashMap benchmark to avoid ConcurrentModificationException since we want to test get & computeIfAbsent.
3. Adding back % math operator to avoid unnecessary double for loop and the % I thought is not something that may mess up the benchmark result.
The latest benchmark result is here (also modified the pr description). And it clearly showed out the performance of concurrentHashMap is better than CopyOnWriteMap. 
```
Benchmark                                     (mapSize)  (times)  (writePerLoops)  Mode  Cnt          Score          Error  Units
ConcurrentMapBenchmark.testConcurrentHashMap        100  1000000            10000  avgt   15   15046454.526 ±    39687.442  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMap           100  1000000            10000  avgt   15  236311312.944 ± 21788514.733  ns/op
```

brandboat committed:
Remove redundant code

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ConcurrentMapBenchmark.java:
```diff
@@ -0,0 +1,92 @@
+        }
+    }
+
+    @Benchmark
```
Could you please `@OperationsPerInvocation(value = times)` to make it show average time of "op"?

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ConcurrentMapBenchmark.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @Benchmark
+    public void testConcurrentHashMap(Blackhole blackhole) {
```
Could you please add benchmark for "entrySet` and `values`? Both methods are used by producer

brandboat replied:
thanks ! now the benchmark result are as belows.

```
# OS: openSUSE Tumbleweed
# JMH version: 1.37
# VM version: JDK 21.0.3, OpenJDK 64-Bit Server VM, 21.0.3+9-LTS (zulu)
# CPU: 13th Gen Intel(R) Core(TM) i9-13980HX

Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt  Score   Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt   15  2.368 ± 0.066  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt   15  4.385 ± 0.013  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt   15  8.895 ± 0.019  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt   15  2.289 ± 0.052  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt   15  2.442 ± 0.072  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt   15  4.664 ± 0.014  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt   15  6.243 ± 0.033  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt   15  2.435 ± 0.174  ns/op
```

brandboat committed:
Address comments

chia7712 commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ConcurrentMapBenchmark.java:
```diff
@@ -0,0 +1,175 @@
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                concurrentHashMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(concurrentHashMap.values());
```
maybe we need to iterate it. the time complexity can be very different

brandboat replied:
Thanks, the benchmark result are as belows

```
Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt      Score     Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt   15  30675.669 ± 305.105  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt   15      2.494 ±   0.024  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt   15      5.134 ±   0.083  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt   15  33012.029 ± 677.201  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt   15  34543.961 ± 398.451  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt   15      2.845 ±   0.041  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt   15      4.218 ±   0.062  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt   15  26139.956 ± 229.991  ns/op
```

chia7712 replied:
It is weird ... why `CopyOnWrite#entrySet` is slower but  `CopyOnWrite#Values` is not ?

brandboat replied:
It is a bug, setup level should be Invocation instead of the default value (Trial), otherwise computeIfAbsent won't work in the iterations other than the first iteration.

brandboat replied:
Now the benchmark result are as belows

```text
Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt      Score      Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt    5  22338.964 ± 3148.123  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt    5     13.175 ±    0.488  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt    5     16.382 ±    1.301  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt    5  19684.342 ± 2892.526  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt    5  21260.254 ±  860.984  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt    5   5436.128 ±  883.432  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt    5   5606.463 ±  962.939  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt    5  16483.967 ± 1072.444  ns/op
```

chia7712 replied:
It seems the cost of `testCopyOnWriteMapGet` is the "write" need to do deep-copy, right? 

Maybe we can revise the use case of `RecordAccumulator#nodeStats`. It updates rarely as the number of nodes won't be changed frequently. Also, it call `get` only.

brandboat replied:
Make sense to me, should I open a JIRA or just create a minor pr for this one ?

chia7712 replied:
> Make sense to me, should I open a JIRA or just create a minor pr for this one ?

yep. BTW, could you please add comments to `CopyOnWriteMap` to describe this benchmark?

brandboat replied:
filed https://issues.apache.org/jira/browse/KAFKA-16942

brandboat committed:
Address comments

brandboat committed:
Address comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

brandboat committed:
Add comments

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

chia7712 reviewed (approved):
LGTM

chia7712 commented:
@brandboat could you please remove `Microbenchmarks` from suppression file?

https://github.com/apache/kafka/blob/6d1f8f8727e8f9c7071b83b7dd9feb840d8d661f/checkstyle/suppressions.xml#L61

brandboat committed:
Address comment

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

brandboat committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-12708

brandboat committed:
Fix code style

chia7712 reviewed (approved):
LGTM
